cppcache/src/statistics/StatisticsTypeImpl.cpp

```diff
@@ -29,35 +29,28 @@ namespace statistics {
 using client::IllegalArgumentException;
 using client::NullPointerException;
 
-StatisticsTypeImpl::StatisticsTypeImpl(std::string nameArg,
-                                       std::string descriptionArg,
-                                       StatisticDescriptor** statsArg,
-                                       int32_t statsLengthArg) {
+StatisticsTypeImpl::StatisticsTypeImpl(
+    std::string nameArg, std::string descriptionArg,
+    std::vector<std::shared_ptr<StatisticDescriptor>> statsArg) {
   if (nameArg.empty()) {
     const char* s = "Cannot have an empty statistics type name";
     throw NullPointerException(s);
   }
-  if (statsArg == nullptr) {
-    const char* s = "Cannot have a null statistic descriptors";
-    throw NullPointerException(s);
-  }
-  if (statsLengthArg > MAX_DESCRIPTORS_PER_TYPE) {
+  if (stats.size() > MAX_DESCRIPTORS_PER_TYPE) {
     throw IllegalArgumentException(
-        "The requested descriptor count " + std::to_string(statsLengthArg) +
+        "The requested descriptor count " + std::to_string(stats.size()) +
         " exceeds the maximum which is " +
         std::to_string(MAX_DESCRIPTORS_PER_TYPE) + ".");
   }
   this->name = nameArg;
   this->description = descriptionArg;
-  this->stats = statsArg;
-  this->statsLength = statsLengthArg;
+  this->stats = std::move(statsArg);
   int32_t intCount = 0;
   int32_t longCount = 0;
   int32_t doubleCount = 0;
-  for (int32_t i = 0; i < this->statsLength; i++) {
+  for (auto stat : stats) {
     // Concrete class required to set the ids only.
-    StatisticDescriptorImpl* sd =
-        dynamic_cast<StatisticDescriptorImpl*>(stats[i]);
+    auto sd = std::dynamic_pointer_cast<StatisticDescriptorImpl>(stat);
     if (sd != nullptr) {
```

Comment:
Small preference for `std::dynamic_pointer_casts`. If used as a conditional just declare in the conditional to reduce scope.
```c++
if (auto sd = std::dynamic_pointer_cast<StatisticDescriptorImpl>(stat)) {
}
```
The nice thing about `std::shared_ptr`, and `std::std_ptr`, is that they decay to `bool` gracefully.

Reply:
Cool!